### ISSUE TITLE
Explore scalability of eg. Class D sizes

### ISSUE THREAD

**Issue by npadmana** [npadmana/DistributedFFT#15]
In the timing runs for the PAW submission, we noticed that the Class D problem sizes started to saturate at about 128 locales (see #14 for graphs).

In part, this may be due to overheads in bulk comm, but I suspect this is also because our algorithm doesn't expose enough parallelism to fully use these locales. 

We do a 1D data decomposition and use a single task to handle each plane separately. For the Class D problem at 128 locales, we have 8 planes per locale so it isn't too surprising that things are starting to saturate. 

There are two possible solutions here --
1. Change our data decomposition to a 2D decomposition (this is what routines like https://github.com/mpip/pfft) do. However, this requires some significant work, so not a high priority.
2. Instead of calling into FFTW to do a single plane at a time, call into FFTW to do a single line at a time, and then run all the lines in parallel. There is an overhead here with nested `forall`s, but I don't have a good sense for how bad that will be. But this should significantly increase the amount of expressed parallelism and if the overheads are small enough, should scale to larger numbers of locales.

Just for definiteness, here's what this might look like --
```chpl
forall iplane in localPlanes {
  forall iline in 0..nlines {
     fft_line(iline);
  }
}
```

@ronawho -- do you expect this to be particularly problematic? If not, we should definitely try it out....

**Comment by ronawho** [comment]
Ah interesting. Nested foralls should perform just fine. If the outermost forall uses all the cores, the inner most forall will be serial (our leader iterators throttle how much parallelism is created to avoid overloading the system)

There could be overhead for calling fftw on smaller chunks (either not as good of cache utilization, fftw setup, whatever else) but I think this is worth trying out.

**Comment by ronawho** [comment]
Let me know if you get a chance to try this out -- I'm happy to rerun scalability testing

**Comment by npadmana** [comment]
The branch `pencilFFT` tries to expose more parallelism by doing a series of 1D transforms instead of 2D transforms.

There is a new config parameter as well `numberOfPlanes`. This defaults to 2, but we might get better performance by bumping this up to a somewhat larger number (say 4 or 6)? As a nice plus, this should also fix the OOM problem, since we limit the number of planes used.

@ronawho -- I haven't had a chance to test on swan since it's down. If you get a chance, it would be interesting to see if this improves the performance on the Class D problems (and doesn't regress on the other cases). Thanks!

**Comment by ronawho** [comment]
Seems to hurt performance for <=128 locales  and improve for 256/512 (but still off from reference)


Size D comparing reference vs. master vs. pencilFFT with `--numberOfPlanes=2`:

| locales | ref   | master | nOP=2 |
| ------- | ----- | ------ | ----- |
|  32     | 14.1s | 14.1s  | 18.8s |
|  64     | 10.45 |  9.5s  | 14.3s |
| 128     |  5.3s |  7.3s  |  9.3s |
| 256     |  2.9s |  6.1s  |  5.8s |
| 512     |  1.5s |  5.6s  |  4.7s |


Size D comparing master vs. pencilFFT with various `--numberOfPlanes` (nOP):

| locales | master | nOP=1 | nOP=2 | nOP=4 | nOP=8 |
| ------- | ------ | ----- | ----- | ----- | ----- |
|  32     | 14.1s  | 18.6s | 19.8s | 19.8s | 18.0s |
|  64     |  9.5s  | 12.9s | 14.3s | 12.7s | 11.3s |
| 128     |  7.3s  | 10.0s |  9.3s |  7.9s |  7.7s |
| 256     |  6.1s  |  6.1s |  5.8s |  5.8s |   N/A |
| 512     |  5.6s  |  4.5s |  4.7s |   N/A |   N/A |

(N/A cases were hitting an assert somewhere -- not sure where yet)

**Comment by npadmana** [comment]
Thanks!

I know the issue with the assert -- in these cases, there aren't enough planes locally stored, and I didn't do a check. That's an easy fix.

I was hoping this simple version would catch up with the reference. Need to stare at this some more.

**Comment by npadmana** [comment]
Just collecting a few random thoughts here -- 

1. For the transposed case, I wrote the copy back into the destination array in Chapel (in the hope that this copy would be multithreaded), see ![here](https://github.com/npadmana/DistributedFFT/blob/dfd19c3459ab84ee16f304ffad58e472c1084096/src/DistributedFFT.chpl#L474). I wonder if going back to a simple memcpy is the better approach.

2. Do we actually need the unaligned flag? It's worth doing a few experiments here to see what the alignments of the different 1D arrays are. That being said, even the `master` version was running with the unaligned flag, so this may be less of an issue.

**Comment by npadmana** [comment]
@ronawho -- I've fixed the bug with numberOfPlanes (I think) --- I don't have a great test case for that on my laptop :-). I've also switched to memcpy's instead of a Chapel copy. 

The default `numberOfPlanes` is now 1, in part motivated by the numbers you had got previously. But this might also depend on details of the problem, so I'll leave this parameter in. We might change the default again.

Any chance you could re-run the Case D benchmark?

**Comment by ronawho** [comment]
Best of 3 runs using d963e2b:

| locales | master | nOP=1 |  nOP=8 |
| ------- | ------ | ----- |  ----- |
|  32     | 14.1s  | 16.5s |  17.6s |
|  64     |  9.5s  | 11.9s |  11.1s |
| 128     |  7.3s  |  8.7s |   7.5s |
| 256     |  6.1s  |  6.3s |   5.8s |
| 512     |  5.6s  |  4.4s |   4.8s |

**Comment by npadmana** [comment]
Hm. I guess that's better, even if not dramatically so.

fcc3f673e52dec78329d makes a tweak to the `evolve` function (so it won't close the gap with master, since this should be applied to both), but it does slightly better on my laptop. I wonder if it gets us closer to the reference case. 

I suspect FFTW is able to make some clever optimizations when running each plane at a time, that we're losing by doing each line at a time. However, we pay the price in amount of parallelism, given the data decomposition.

**Comment by npadmana** [comment]
Here are some thoughts about closing the gap both with `master` and `ref`. 

For `master` : the YZ transform piece is done plane-by-plane, so we do `NX/numLocales` calls to FFTW. This, of course, restricts parallelism when `numLocales` is close to `NX`, since the FFTW calls are serial.

For `pencilFFT`: we do a series of 1D transforms, so we end up doing `NX*(NY + NZ)/numLocales` calls to FFTW. While this exposes more parallelism, we pay a price on overheads. Now, we could try to block up these FFTW calls to minimize this. Unfortunately, I think that this also means that the parallelism must be a lot more explicit (`coforall` instead of `forall`). 

The simple solution here would be a switch that goes from 2D to 1D transforms when the number of planes per locale drops below some threshold.

For the X transform piece, I think there are two pieces. The first is again parallelism vs FFTW overheads, just as in the YZ case. But also, the amount of data being communicated relative to the communication overheads is going down, so we're paying a penalty here (relative to the reference, not master).

**Comment by npadmana** [comment]
Here are some to dos before we put this to bed:

- [ ] Look at the FFT timers output for Case D on 32 locales, just to make sure we understand where we're losing time.
- [ ] Run with the version of `BlockDist` from #4 to see if that gets us closer to the reference.
- [ ] Could we run the FFT calls on blocks? How hard would this be?

**Comment by npadmana** [comment]
Summary -- the most significant slowdown is in the X transform, so that's probably where we should focus to close the gap with master.

```
Raw information below --
For Master:
-----------

8 locales
---------
Elapsed time :    35.3779
MFLOPS : 253371.3358
--------- Timings ---------------
Time for X :      26.65
Time for YZ :       5.43
Time for Execute :     775.41
Time for Comms :    7191.66
---------------------------------


16 locales
----------
Elapsed time :    21.8520
MFLOPS : 410201.6201
--------- Timings ---------------
Time for X :      17.08
Time for YZ :       3.36
Time for Execute :     765.45
Time for Comms :    7828.23
---------------------------------


32 locales
----------
Elapsed time :    14.8388
MFLOPS : 604073.3973
--------- Timings ---------------
Time for X :      10.67
Time for YZ :       2.51
Time for Execute :     685.06
Time for Comms :    7560.69
---------------------------------

For PencilFFT
-------------
We are not tracking Execute time, due to overheads here...

8 locales
---------
Elapsed time :    47.0496
MFLOPS : 190516.8664

--------- Timings ---------------
Time for X :      37.36
Time for YZ :       6.75
Time for Execute :       0.00
Time for Comms :     269.61
---------------------------------

8 locales : nOP = 2
-------------------
Elapsed time :    42.2609
MFLOPS : 212104.8618
--------- Timings ---------------
Time for X :      32.41
Time for YZ :       6.75
Time for Execute :       0.00
Time for Comms :     374.76
---------------------------------

8 locales : nOP = 4
-------------------
Elapsed time :    44.3093
MFLOPS : 202299.4289
--------- Timings ---------------
Time for X :      34.62
Time for YZ :       6.74
Time for Execute :       0.00
Time for Comms :     591.63
---------------------------------

16 locales
----------
Elapsed time :    27.0850
MFLOPS : 330948.9705
--------- Timings ---------------
Time for X :      21.52
Time for YZ :       4.18
Time for Execute :       0.00
Time for Comms :     309.26
---------------------------------

32 locales
----------
Elapsed time :    16.5527
MFLOPS : 541529.0938
--------- Timings ---------------
Time for X :      11.97
Time for YZ :       2.78
Time for Execute :       0.00
Time for Comms :     347.19
---------------------------------
```

**Comment by ronawho** [comment]
Sorry I got busy with the release, but I should have some time to dig now.

> fcc3f67 makes a tweak to the evolve function (so it won't close the gap with master, since this should be applied to both), but it does slightly better on my laptop. I wonder if it gets us closer to the reference case.

I would guess the backend is already optimizing that, but it certainly can't hurt. 

>  Run with the version of BlockDist from #4 to see if that gets us closer to the reference.

Yeah, this is the one I plan to run with for right now. I think there is a decent amount we can do to optimize our comm. I think we should be able to stuff our metadata into a single data structure and issue a single GET instead of multiple smaller GETs. I also think we can optimize our strided transfers to actually use non-blocking comm (disabled in https://github.com/chapel-lang/chapel/pull/13740 because of some hangs we were seeing)

> Could we run the FFT calls on blocks? How hard would this be?

I'm pretty slow to read fortran, but is this what the reference version does? It looks like they have several different strategies for how to block/transpose but it's not obvious to me when they choose what yet.

**Comment by npadmana** [comment]
Eliot -- just a quick update and a question. I think this branch has now largely caught up with master, and hopefully should scale better as well.

For now, should I just use the default `BlockDist` and wait until your optimizations are in? Or is it worth making a (renamed) copy of it and play with it in this repo?

**Comment by ronawho** [comment]
Hmm, not sure which route makes the most sense. Let me dig for today and get back to you tomorrow?

**Comment by npadmana** [comment]
Some preliminary timings (I didn't bother averaging, so add +/- 0.1 s to any number). The numbers in parens are the (X/Y) individual timings (X has communication, Y does not). These are all nOP=1 and using 590f2b76 .

| locales | master |  pencilFFT | 
|---|---|---|
| 8 | 35.3 (26.6/5.4)  | 42 (32.1/6.7) |
| 16 | 21.8 (17.1/3.4) | 24.5 (18.7/4.2) | 
| 32 | 14.8 (10.7/2.5) | 14.9 (10.3/2.8) |

**Comment by ronawho** [comment]
FWIW the reference switches from a 1D to a 2D layout depending on the number of nodes (or maybe ranks) being used. Here's a table showing when the transition is made:

| nodes | D  | E  | F  |
| ----- | -- | -- | -- |
|   8   | 1D | 1D | -  |
|  16   | 1D | 1D | -  |
|  32   | 1D | 1D | -  |
|  64   | 2D | 1D | 1D |
| 128   | 2D | 2D | 1D |
| 256   | 2D | 2D | 2D |
| 512   | 2D | 2D | 2D |

And here's a graph of the reference version for size D/E (the perf when the 1D -> 2D switch happens is interesting):

![npb-ft-ref-layout-inflection](https://user-images.githubusercontent.com/1588337/64982580-7e730900-d88c-11e9-9591-2180420594b9.png)

(The reference versions prints the layout type, so I just grabbed that from the runs I did previously -- logs are [here](https://github.com/npadmana/DistributedFFT/tree/master/runs/ronawho-2019-08-15-NPB-FT/cray-xc/graphs/raw-data). The graph is from the same data as that and https://github.com/npadmana/DistributedFFT/pull/14, I just merged reference D/E onto a single graph)

**Comment by ronawho** [comment]
>> For now, should I just use the default BlockDist and wait until your optimizations are in? Or is it worth making a (renamed) copy of it and play with it in this repo?

> Hmm, not sure which route makes the most sense. Let me dig for today and get back to you tomorrow?

Took me longer than a day, but my current prototype is available https://github.com/chapel-lang/chapel/compare/master...ronawho:non-blocking-bulk-comm

This is an improvement over previous attempts in https://github.com/npadmana/DistributedFFT/issues/4#issuecomment-522107502. Basically it's creating additional tasks on the same thread as a hacky way to get "non-blocking" comm. Ideally, we would do this without creating additional tasks, but this was easiest for now. This has pretty good performance improvements at higher node counts where we're not fully saturating all cores. I suspect that this will not help as much when all cores are busy, but I'll check with the pencilFFT. Here are some graphs using PrgEnv-intel comparing chapel master to the non-blocking-bulk-comm branch for the current DistributedFFT master(bf40d24):

![npb-ft-D-perf](https://user-images.githubusercontent.com/1588337/64988023-88026e00-d898-11e9-9304-c42b7f186487.png)

![npb-ft-E-perf](https://user-images.githubusercontent.com/1588337/64988028-8a64c800-d898-11e9-823c-00ea034cbdb8.png)

![npb-ft-F-perf](https://user-images.githubusercontent.com/1588337/64988036-8df84f00-d898-11e9-88b5-361b48017606.png)

While this helps, we're still pretty far off for size D at 256/512 nodes. I suspect that's because of the data layout differences.

@npadmana do you know if the pencilFFT branch is effectively the reference 2D data layout?

(Raw times and other info can be found at https://github.com/ronawho/DistributedFFT/tree/nb-comm-timings/runs/ronawho-2019-09-16-NPB-FT-nb-comm/cray-xc)

**Comment by npadmana** [comment]
@ronawho -- No, pencilFFT doesn't do a 2D FFT, but it does try to expose a little more parallelism in the FFT stage. It would certainly be interesting to see how this new `BlockDist` performs with that case. We're no longer spawning a series of plane GETs, and so parallelizing the GET might help. It might also help at the low node counts, where the current GETs aren't very parallel.

I don't expect us to be able to keep scaling up node counts with the current code, but doing a 2D layout is a bunch more work.

**Comment by npadmana** [comment]
It's worth noting that the 2D distribution is much more critical for the MPI reference, since the data is distributed across ranks, and one would quickly saturate a 1D distribution. For the same core count, we end up having more flexibility.

That being said, the MPI version probably has better memory affinity (if I understand how affinity works).

**Comment by ronawho** [comment]
Ok, here are results for pencilFFT w/ and w/o the non-blocking comm. These graphs are pretty busy:

Size D:
-------

![npb-ft-D-perf](https://user-images.githubusercontent.com/1588337/64995163-7fb22f00-d8a8-11e9-841b-e4da5fb206ab.png)

pencil w/ non-blocking comm is pretty good at high node counts and just loses a little at lower node counts. Just pencil is better than master at high node counts, but suffers at lower counts. 

Size E:
-------

![npb-ft-E-perf](https://user-images.githubusercontent.com/1588337/64995423-2eef0600-d8a9-11e9-88a9-14ed5397da26.png)

Just pencil really suffers for size E. However, pencil w/ non-blocking comm does really well at 256/512 nodes

---

Raw time tables:

<details>

Size D:
-------

| nodes | ref    | master | NB-comm | pencil | pencil NB |
| ----- | ------ | ------ | ------- | ------ | --------- |
|  32   |  13.0s |  14.7s |  13.8s  |  14.4s |  15.2s    |
|  64   |  10.7s |   9.6s |   8.4s  |  11.8s |   8.7s    |
| 128   |   5.4s |   7.6s |   5.5s  |   9.8s |   5.1s    |
| 256   |   2.9s |   6.2s |   4.2s  |   7.0s |   3.3s    |
| 512   |   1.5s |   5.6s |   3.4s  |   4.3s |   2.3s    |

Size E:
-------

| nodes | ref    | master | NB-comm | pencil | pencil NB |
| ----- | ------ | ------ | ------- | ------ | --------- |
|  64   |  60.1s |  63.0s |  59.5s  | 142.7s |  93.5s    |
| 128   |  52.01 |  41.6s |  38.0s  | 113.1s |  49.1s    |
| 256   |  24.9s |  30.1s |  25.0s  |  86.8s |  25.7s    |
| 512   |  12.8s |  25.4  |  18.7s  |  53.5s |  13.4s    |

</details>

**Comment by ronawho** [comment]
Here's my current thoughts:

The non-blocking comm seems to be a win, That's something I will continue to run with and see if I can get that on Chapel-lang master.

Pencil performance seems to suffer (sometimes significantly) at lower node counts, but can definitely help at high node counts. I'm not sure this is worth trying to merge into master at this time. Maybe if we do, we only use it once we know core saturation will be low?

At a higher level though, I'm not sure I care that much about making NPB-ft performance at higher node counts when we're no longer saturating all the cores. It would be great if we were on par, but I think we have a good explanation for why we lose that isn't because of something inherent in Chapel or DistributedFFT, but more of a missing optimization. If you have other thoughts on improving performance I'm happy to do perf runs, but I'm wondering if we should shift our attention to optimizing other benchmarks/applications. Thoughts?

**Comment by npadmana** [comment]
@ronawho -- I think I largely agree. I scribbled down some notes on all of this, see below...

Summary
-------

* The low-node count cases for `pencilFFT` are probably losing because we aren't fully saturating the
  the communication, compared to `master`.
* `pencilFFT` has a significant memory advantage over `master`, which might make it advantageous
  if we can close the gap a little more at the low node count case.
* I agree that we don't want to spend too much more time optimizing this case, except the previous point. I don't think we should obsess about matching reference (although the case E case with pencilFFT and non-blocking comm is very cool!).


Baseline case
-------------

Just for reference, a good typical problem size is Case D on 8 locales - I suspect problems of that size will be our initial workhorse problems. However, we definitely are interested in running larger problems, so it's worth keeping track of scaling out to Case F.

Comparison of algorithms : master
---------------------------------

It's worth comparing the two approaches used. The key difference is in the `X` transform communication
(there are some differences/inefficiencies in the actual FFT, but these are a small fraction of the time).

On `master`, the `X` transform looks like :

```chpl
forall iy in localY {
  // myplane is task-local
  var myplane = arr[..,iy,..];

  // FFT myplane in the X direction with
  // in this task

  // copy to destination - local operation for the benchmark
}
```

* This saturates the cores when `localY.size` (which is approximately Nx/numLocales) is greater than numCores. For large numbers of cores, we're not using most of the machine.
* This uses ~ numCores extra planes as work space. This can become significant in terms of memory usage.
* There isn't much task creation here, except for the beginning.

Comparison of algorithms : pencilFFT
------------------------------------

The pencilFFT algorithm, with a default of `numberOfPlanes=1` looks like

```chpl
for iy in localY {
  var myplane = arr[..,iy,..]; // The GET

  forall iz in zRange {
    // FFT myplane[..,iz]
  }

  // Copy to destination - local operation for the benchmark
}
```

* For low node counts, the "GET" isn't using the all cores. I believe this is where pencilFFT is losing to master.
* Since the FFTs are all done line-by-line, there is a lot of parallelism here.

One possibility to improve the scaling of the GET at low node counts is 

```chpl
forall ix in xRange do myplane[ix,iy,..] = arr[ix,iy,..];
```
Note that `arr[ix,iy,..]` is both local and contiguous, so the actual data fetch should require one GET. That being said, there will be some overhead sending the metadata over. Or we could try something along the remote `memcpy` lines I was talking about.

We could also pull multiple planes over at a time. The current implementation of this doesn't quite seem to do the trick, but we could try a variant of this approach.

**Comment by ronawho** [comment]
Ah thanks, that algorithm comparison was really useful! For plain pencil we're almost certainly not saturating the network bandwidth, so it makes sense that larger problem sizes would be especially hurt since there's so much more comm required.

Ok, with that in mind here's a really quick attempt to get better bandwidth saturation -- https://github.com/ronawho/chapel/tree/non-blocking-bulk-comm-2. It's the nb-comm branch, but instead of spawning all tasks to the same thread, we're now spreading them across (I didn't realize pencil's comm occurred in a serial section.)

With that size D and E performance looks really good at higher node counts -- size D is even beating the reference.  64 node E is a little off still, and I would guess performance is worse on smaller node counts too, but overall looks promising.

![npb-ft-D-perf](https://user-images.githubusercontent.com/1588337/65008439-5fe62f80-d8d7-11e9-82d7-ac2a02997c9a.png)

![npb-ft-E-perf](https://user-images.githubusercontent.com/1588337/65008443-62488980-d8d7-11e9-97fb-1522d3bfdb8f.png)

(These graphs are reference vs. fft-master with chapel nb (spawn comm tasks to same thread) vs fft-pencil with chapel nb2 (spawn comm tasks to different threads))

I'll keep this in mind while optimizing bulk-comm -- I think I can get something that should perform well for either fft-master or fft-pencil

**Comment by npadmana** [comment]
Wow. That's really quite nice.....

A question and a request. Is Chapel limiting to 36 cores/locale, or is that only the reference version? I'm wondering if we're doing a 1-1 comparison here.

The request -- in your timing runs, could you run case D for 8 and 16 as well, and possibly run case E for 32? I'm just curious as to the price we're paying at low node counts for `master` vs `pencilFFT`.

**Comment by npadmana** [comment]
Another question -- the line
```chpl
const numTasks = min(activeLocs.domain.size, here.maxTaskPar);
```
would seem to suggest that we get rate limited again when numCores > numLocales. Is there something we can do to better saturate in this case?

**Comment by ronawho** [comment]
> Is Chapel limiting to 36 cores/locale, or is that only the reference version? I'm wondering if we're doing a 1-1 comparison here.

The larger machine I'm running on (Crystal) has 36-core Broadwell CPUs. Swan has 44-core Broadwell CPUs, but we can only do up to 64 node runs there. The reference version limits to a power of 2, so we end up only using 32 of the 36 cores. In earlier experiments, I did on swan with limiting Chapel to 32 cores I did not see much of a performance difference (I suspect 32 cores is enough to both saturate the local memory bandwidth and the NIC.) That said I can do some more 32-core Chapel runs tomorrow to get a more apples-to-apples comparison.

> in your timing runs, could you run case D for 8 and 16 as well, and possibly run case E for 32? I'm just curious as to the price we're paying at low node counts for master vs pencilFFT

Yup, I'll gather some lower node count runs tomorrow (they're slower to run so I omitted them just to speed up gathering results while experimenting)

**Comment by ronawho** [comment]
> `const numTasks = min(activeLocs.domain.size, here.maxTaskPar);`
> would seem to suggest that we get rate limited again when numCores > numLocales. Is there something we can do to better saturate in this case?

I played with `min(activeLocs.domain.size, 4*here.maxTaskPar);` to try to get some "non-blocking" comm overlap within each thread, but it performed a little worse. I'm not sure why yet though. What I think we really want is a top-level forall that throttles the number of tasks created, and then within the forall initiate non-blocking comm. That should give us the best of both worlds and work well for fft-master and fft-pencil.

We still have to be careful that we're spreading the load though. We don't want all remote nodes hitting specific nodes in lockstep or we end up introducing congestion (i.e we don't want all remote nodes hitting node 0 all at once.) My best guess is that the `4*here.maxTaskPar` introduced congestion somewhere, but I need to dig more.

**Comment by npadmana** [comment]
Maybe I'm missing something here, but the case I was worried about was when `activeLocs.domain.size` < `here.maxTaskPar` (i.e. the low node-count limit). In this case, it looks like the number of tasks started is determined by `numLocales`....

However, since the data are non-contiguous, I was wondering if multiple GETs could be issued in parallel.

**Comment by ronawho** [comment]
Oh, I misunderstood. Yeah, good point --  we won't fully saturate comm when `numLocales < numCores`. Hmm, let me think about that. Today we just call into a strided comm interface, but we may be able to parallelize something in there. 

(The strided comm used to do non-blocking comm internally, but that was disabled in https://github.com/chapel-lang/chapel/pull/13740 -- fixing that so we use non-blocking comm again would help, but increasing parallelism would almost certainly help more.)

**Comment by ronawho** [comment]
Here's what I see for small scale size D comparing reference vs. master vs. pencilFFT w/ the non-blocking-bulk-comm-2 branch:

| nodes | ref    | master | pencil nb-comm-2 |
| ----- | ------ | ------ | ---------------- |
|  1    | 262.2s | 158.3s | 352.3s           |
|  2    | 156.5s |  85.1s | 137.4s           |
|  4    |  93.8s |  57.9s |  72.8s           |
|  8    |  51.0s |  36.1s |  39.9s           |
| 16    |  25.4s |  20.5s |  22.6s           |
| 32    |  14.1s |  14.0s |  14.1s           |

Not surprisingly, performance suffers compared to master until we get to the point where numLocales is close to numCores (since as you pointed out we're not saturating the cores with comm here.)

**Comment by npadmana** [comment]
So, I did try parallelizing the GET with a `forall`; it made things a little worse for case D on 8 locales. My initial suspicion was that the overheads (for the metadata) was causing an issue here. It could also that we're hitting the same nodes in sync and that is causing an issue.

I did try dropping down to a low-level call to `chpl_comm_array_get`. That did a little better, but still not as well as a single GET.

**Comment by ronawho** [comment]
Mind putting that work on a branch for me to experiment with?

**Comment by npadmana** [comment]
Sure, give me a few minutes....

**Comment by npadmana** [comment]
`pencilFFT-1` it is.

All my experimentation is in `doFFT_X_Transposed`, around line 471. But I really don't know what I'm doing here, so I'm hopeful that you'll find something!

**Comment by ronawho** [comment]
Huh.. yeah, I'm not sure why that hurt performance for 8 locales. I've tried a few different things, but nothing is jumping out at me. I'll dig more later. FWIW this helped a lot at lower node counts, and it looks great when you just compare it to the reference:

| nodes | reference | pencilFFT-1 |
| ----- | --------- | ----------- |
|   1   | 259.2s    | 165.3s      |
|   2   | 156.3s    | 119.3s      |
|   4   |  90.9s    |  76.8s      |
|   8   |  49.7s    |  46.0s      |
|  16   |  25.0s    |  23.5s      |
|  32   |  12.6s    |  14.0s      |
|  64   |  10.5s    |   7.5s      |
| 128   |   5.2s    |   4.2s      |
| 256   |   2.7s    |   2.4s      |
| 512   |   1.5s    |   1.5s      |


Wih master:

<details>

Adding in master, we can see that master wins up to 32 nodes, but after that, it's pretty far off:

| nodes | master | pencilFFT-1 |
| ----- | ------ | ----------- |
|   1   | 155.5s | 165.3s      |
|   2   |  84.7s | 119.3s      |
|   4   |  58.1s |  76.8s      |
|   8   |  36.2s |  46.0s      |
|  16   |  20.5s |  23.5s      |
|  32   |  14.2s |  14.0s      |
|  64   |   9.4s |   7.5s      |
| 128   |   7.1s |   4.2s      |
| 256   |   5.9s |   2.4s      |
| 512   |   5.6s |   1.5s      |

</details>

(I'm timing out looking for this week, but I'll try to pick this up next week. To start I want to figure out where the performance loss between master and pencilFFT-1 is (comm, compute, memcpy) and I'll work from there.)

**Comment by npadmana** [comment]
(From my phone)

Interesting. Yes, that looks quite nice. Is this with the non blocking
comm, or just Chapel master?

I’m probably going to try and focus on the downstream simulation code to
get a working version by next week so that we can clean up our abstract.
Will you be able to make it out for PAW?

**Comment by ronawho** [comment]
> Interesting. Yes, that looks quite nice. Is this with the non blocking comm, or just Chapel master?

Chapel master (since we're manually doing comm, it doesn't matter though)

> I’m probably going to try and focus on the downstream simulation code to get a working version by next week so that we can clean up our abstract.

Makes sense to me. I think FT is an interesting comparison and a good indicator that we can get good perf/scalability, but I think the actual simulation is what we should switch our attention. Feel free to ping me if you want some help optimizing the real code.

> Will you be able to make it out for PAW?

Unfortunately, I don't think so but Michael and Brad will be there.

**Comment by npadmana** [comment]
@ronawho -- if you get a chance, could you run the perf tests for `pencilFFT-2`? For low node counts on swan, I see somewhat better performance than `pencilFFT-1`, but I'm not sure how it will behave at large node counts.....

**Comment by ronawho** [comment]
I have a [pencilFFT-1-e branch](https://github.com/npadmana/DistributedFFT/compare/pencilFFT-1...ronawho:pencilFFT-1-e?w=1) that slightly improves pencilFFT-1 by fusing the memcpy and comm loops:

Size D:
-------

| nodes | master | pencilFFT-1 | pencilFFT-1-e |
| ----- | ------ | ----------- | ------------- |
|   1   | 155.5s | 165.3s      | 159.7s        |
|   2   |  84.7s | 119.3s      | 109.9s        |
|   4   |  58.1s |  76.8s      |  71.9s        |
|   8   |  36.2s |  46.0s      |  41.6s        |
|  16   |  20.5s |  23.5s      |  21.6s        |
|  32   |  14.2s |  14.0s      |  13.2s        |
|  64   |   9.4s |   7.5s      |   7.0s        |
| 128   |   7.1s |   4.2s      |   4.0s        |
| 256   |   5.9s |   2.4s      |   2.3s        |
| 512   |   5.6s |   1.5s      |   1.5s        |

**Comment by ronawho** [comment]
Here's the results for your pencilFFT-2 branch (I think it's better than pencilFFT-1 up to 8 locales and worse after that):

Size D:
-------

| nodes | master | pencilFFT-1 | pencilFFT-2 |
| ----- | ------ | ----------- | ----------- |
|   1   | 155.5s | 165.3s      | 156.4s      |
|   2   |  84.7s | 119.3s      | 114.0s      |
|   4   |  58.1s |  76.8s      |  78.8s      |
|   8   |  36.2s |  46.0s      |  43.5s      |
|  16   |  20.5s |  23.5s      |  25.4s      |
|  32   |  14.2s |  14.0s      |  16.9s      |
|  64   |   9.4s |   7.5s      |   9.5s      |
| 128   |   7.1s |   4.2s      |   6.3s      |
| 256   |   5.9s |   2.4s      |   4.5s      |
| 512   |   5.6s |   1.5s      |   2.6s      |

**Comment by npadmana** [comment]
Interesting optimization in `pencilFFT-1-e` -- I'm guessing this is the overhead in going in and out of forall loops?

`pencilFFT-2` was aiming to do something similar by just directly copying into the destination array. There was a small issue with the last version -- it was running the FFT plan creation more often than necessary. Now fixed.

I'm guessing part of the remaining inefficiencies are the nested foralls. We might also win if we could start the next set of gets while the FFTs are executing (in the current version, there are no race conditions) (i.e. we don't need to wait for the FFTs to get done).

**Comment by npadmana** [comment]
Just another thought — pencilFFT-2 should be doing strictly less work than
FFT-1 (no memcpys), and in the limit that the outer forall is just serial,
should be otherwise identical.

So I don’t really have a good model in my head for why it is losing
relative to FFT-1.

**Comment by ronawho** [comment]
> Interesting optimization in pencilFFT-1-e -- I'm guessing this is the overhead in going in and out of forall loops?

Creating tasks is relatively cheap (should be unmeasurably fast here.) I think it's more likely that we're able to get a little bit of overlap with the comm and the memcpy 

<details>
Aries has a Block Transfer Engine (BTE) that we use for comm over 4KB. From the Aries docs: `The BTE unit supports up to four concurrent block transfers. If the transfer is large, then a single active channel is sufficient to sustain full bandwidth. The ability to pipeline transfers on multiple BTE channels ensures that full bandwidth can be maintained at smaller transfer sizes`. So I think when we're doing bulk comm some of the tasks are waiting to initiate a transfer and the tasks will get slightly offset in timings so some tasks will be doing memcpy's while the others are doing comm.
</details>


> pencilFFT-2 was aiming to do something similar by just directly copying into the destination array. There was a small issue with the last version -- it was running the FFT plan creation more often than necessary. Now fixed.

Rerunning now. 

> I'm guessing part of the remaining inefficiencies are the nested foralls. We might also win if we could start the next set of gets while the FFTs are executing (in the current version, there are no race conditions) (i.e. we don't need to wait for the FFTs to get done).

Yeah, I was thinking about this too. Currently, we don't expose the chpl_comm_get_nb primitive, but we can probably fake non-blocking comm by initiating in a different task and using an atomic to synchronize (I think we played with something like that earlier.) I can also take a look at how much work it would be to expose the non-blocking interface.

**Comment by ronawho** [comment]
Updated pencilFFT-2 (07662c4) results:

Size D:
-------

| nodes | master | pencilFFT-1 | pencilFFT-2 |
| ----- | ------ | ----------- | ----------- |
|   1   | 155.5s | 165.3s      | 134.5s      |
|   2   |  84.7s | 119.3s      |  93.2s      |
|   4   |  58.1s |  76.8s      |  70.3s      |
|   8   |  36.2s |  46.0s      |  41.4s      |
|  16   |  20.5s |  23.5s      |  25.3s      |
|  32   |  14.2s |  14.0s      |  15.3s      |
|  64   |   9.4s |   7.5s      |   9.3s      |
| 128   |   7.1s |   4.2s      |   6.5s      |
| 256   |   5.9s |   2.4s      |   4.0s      |
| 512   |   5.6s |   1.5s      |   2.4s      |

**Comment by ronawho** [comment]
I know most of the time is spent doing the X transform, but I'm curious if the YZ transform has limited parallelism at higher locale counts

**Comment by npadmana** [comment]
No, I rewrote it to do the YZ piece as a series of 1D transforms so that
should scale reasonably well. There’s overhead from repeated FFT calls, but
I think that is small/amortized.

**Comment by npadmana** [comment]
—timeTrackFFT should show the time in both YZ and X.

**Comment by ronawho** [comment]
My current theory for why pencilFFT-1 loses at lower node counts is because of poor memory affinity. We have something like:

```chpl
var myplane : [{xRange, 0..0, zRange}] T;

forall iz in zRange with (ref plan_x) {
  var elt = c_ptrTo(myplane[x0, 0, iz]);
  plan_x.execute(elt, elt);
}
```

Where the `myplane` declaration is setting up numa-affinity, but then we're not spawning tasks in the same way for the fft portion. So I think what's happening is some tasks are doing fft operations on data that lives on a different numa domain so there's overhead. Once we get enough locales the benefits of increased parallelism outweigh the numa overhead. 

(Currently, we use a process per node and rely on first-touch to set up affinity. Over the past few months we have been discussing if we should look at a process per numa-domain instead.)

**Comment by npadmana** [comment]
That makes sense. Do you understand the difference between pencilFFT-1 and
-2? There is less memory traffic in 2 (no separate memcpy), which is why I
think we win on low node counts. I wonder at large node counts, we’re
getting hurt by cache.

myplane does fit into L3 cache, or comes close to.

**Comment by ronawho** [comment]
Hmm, just looking at the code I'm not sure. 

I'll have to do some timing runs and dig a little to get a sense of what's going on.

**Comment by npadmana** [comment]
@ronawho -- 7b540e039e1a4 is my best version for `pencilFFT-2` -- it shouldn't be much worse than `master`. Could you run the `D` timings on this? If we're mostly happy with this, then I'll set up a switch that goes between `pencilFFT-1` and `pencilFFT-2`.

**Comment by ronawho** [comment]
I don't see much of a difference between 07662c4 and 7b540e0 (maybe 2-3% improvement)

Oh, now I see the difference between pencilFFT-1 and pencilFFT-2(both at 07662c4 and 7b540e0) -- Making the yChunk loop a parallel forall is what's hurting at higher locale counts. What's probably happening is the forall is creating let's say 2 tasks, and then the inner foralls are creating numCores/2 tasks, but with our current scheduler this isn't getting distributed across all the cores.   If you switch back to a for loop (revert 7b540e039e1a4b5f81c6831d5c221d6718cb00b0) performance is better with pencilFFT-2. I gather full numbers when I have access to the machine I've been running on again.

**Comment by npadmana** [comment]
We understand the lack of scalability of our original code and we're seeing good scaling with many of our new branches. So closing this as done.....
